services/ingestion/block_tracking_subscriber.go

```diff
@@ -0,0 +1,277 @@
+// as well as context cancellations.
+func (r *RPCBlockTrackingSubscriber) subscribe(ctx context.Context, height uint64) <-chan models.BlockEvents {
+	eventsChan := make(chan models.BlockEvents)
+
+	blockHeadersChan, errChan, err := r.client.SubscribeBlockHeadersFromStartHeight(
+		ctx,
+		height,
+		flow.BlockStatusFinalized,
+	)
+	if err != nil {
+		eventsChan <- models.NewBlockEventsError(
+			fmt.Errorf(
+				"failed to subscribe for finalized block headers on height: %d, with: %w",
+				height,
+				err,
+			),
+		)
+		return eventsChan
+	}
+	lastReceivedHeight := height
+
+	go func() {
+		defer func() {
+			close(eventsChan)
+		}()
+
+		for ctx.Err() == nil {
+			select {
+			case <-ctx.Done():
+				r.logger.Info().Msg("event ingestion received done signal")
+				return
+
+			case blockHeader, ok := <-blockHeadersChan:
+				if !ok {
+					var err error
+					err = errs.ErrDisconnected
+					if ctx.Err() != nil {
+						err = ctx.Err()
+					}
+					eventsChan <- models.NewBlockEventsError(err)
+					return
+				}
+
+				blockEvents, err := r.evmEventsForHeight(ctx, blockHeader.Height)
+				if err != nil {
+					eventsChan <- models.NewBlockEventsError(err)
+					return
+				}
+
+				evmEvents := models.NewSingleBlockEvents(blockEvents)
+				// if events contain an error, or we are in a recovery mode
+				if evmEvents.Err != nil || r.recovery {
+					evmEvents = r.recover(ctx, blockEvents, evmEvents.Err)
+					// if we are still in recovery go to the next event
+					if r.recovery {
+						continue
+					}
+				}
+
+				for _, evt := range blockEvents.Events {
+					r.keyLock.UnlockKey(evt.TransactionID)
+				}
+				r.keyLock.Notify(blockHeader.Height)
+				lastReceivedHeight = blockHeader.Height
+
+				eventsChan <- evmEvents
+
+			case err, ok := <-errChan:
+				if !ok {
+					var err error
+					err = errs.ErrDisconnected
+					if ctx.Err() != nil {
+						err = ctx.Err()
+					}
+					eventsChan <- models.NewBlockEventsError(err)
+					return
+				}
+
+				if strings.Contains(errors.Unwrap(err).Error(), "DeadlineExceeded") ||
+					strings.Contains(errors.Unwrap(err).Error(), "unexpected EOF") {
+					blockHeadersChan, errChan, err = r.client.SubscribeBlockHeadersFromStartHeight(
+						ctx,
+						lastReceivedHeight+1,
+						flow.BlockStatusFinalized,
+					)
+					if err != nil {
+						eventsChan <- models.NewBlockEventsError(
+							fmt.Errorf(
+								"failed to subscribe for finalized block headers on height: %d, with: %w",
+								height,
+								err,
+							),
+						)
+						return
+					}
+				} else {
+					eventsChan <- models.NewBlockEventsError(fmt.Errorf("%w: %w", errs.ErrDisconnected, err))
+					return
+				}
+			}
+		}
+	}()
+
+	return eventsChan
+}
+
+func (r *RPCBlockTrackingSubscriber) evmEventsForHeight(
+	ctx context.Context,
+	height uint64,
+) (flow.BlockEvents, error) {
+	eventTypes := blocksFilter(r.chain).EventTypes
+	evmBlockEvent := eventTypes[0]
+
+	evts, err := r.client.GetEventsForHeightRange(
```

Comment:
an important detail from the issue that I missed in the last review is that this needs to use `GetEventsForBlockIDs` instead of the height range version. The range endpoint will restrict the request to _sealed_ blocks, which defeats the purpose of this change. `GetEventsForBlockIDs` will search any blocks you provide allowing use to query finalized blocks as well.